server/util/db/db.go

```diff
@@ -119,12 +121,80 @@ func maybeRunMigrations(dialect string, gdb *gorm.DB) error {
 	return nil
 }
 
-func openDB(dialect string, args ...interface{}) (*gorm.DB, error) {
-	gdb, err := gorm.Open(dialect, args...)
+// instrumentGORM adds GORM callbacks that populate query metrics.
+func instrumentGORM(gdb *gorm.DB) {
+	// Add callback that runs before other callbacks that records when the operation began.
+	// We use this to calculate how long a query takes to run.
+	recordStartTime := func(db *gorm.DB) {
+		if db.DryRun || db.Statement == nil {
+			return
+		}
+		db.Statement.Settings.Store(gormStmtStartTimeKey, time.Now())
+	}
+	gdb.Callback().Create().Before("*").Register("buildbuddy:record_op_start_time", recordStartTime)
+	gdb.Callback().Delete().Before("*").Register("buildbuddy:record_op_start_time", recordStartTime)
+	gdb.Callback().Query().Before("*").Register("buildbuddy:record_op_start_time", recordStartTime)
+	gdb.Callback().Raw().Before("*").Register("buildbuddy:record_op_start_time", recordStartTime)
+	gdb.Callback().Row().Before("*").Register("buildbuddy:record_op_start_time", recordStartTime)
+	gdb.Callback().Update().Before("*").Register("buildbuddy:record_op_start_time", recordStartTime)
+
+	// Add callback that runs after other callbacks that records executed queries and their durations.
+	recordMetrics := func(db *gorm.DB) {
+		if db.DryRun || db.Statement == nil {
+			return
+		}
+		labels := prometheus.Labels{
+			metrics.SQLQueryTemplateLabel: db.Statement.SQL.String(),
+		}
+		v, _ := db.Statement.Settings.LoadAndDelete(gormStmtStartTimeKey)
+		if opStartTime, ok := v.(time.Time); ok {
+			metrics.SQLQueryDurationUsec.With(labels).Observe(float64(time.Now().Sub(opStartTime).Microseconds()))
+		}
+		// Ignore "record not found" errors as they don't generally indicate a
+		// problem with the server.
+		if db.Error != nil && !errors.Is(db.Error, gorm.ErrRecordNotFound) {
+			metrics.SQLErrorCount.With(labels).Inc()
+		}
+	}
+	gdb.Callback().Create().After("*").Register("buildbuddy:record_metrics", recordMetrics)
+	gdb.Callback().Delete().After("*").Register("buildbuddy:record_metrics", recordMetrics)
+	gdb.Callback().Query().After("*").Register("buildbuddy:record_metrics", recordMetrics)
+	gdb.Callback().Raw().After("*").Register("buildbuddy:record_metrics", recordMetrics)
+	gdb.Callback().Row().After("*").Register("buildbuddy:record_metrics", recordMetrics)
+	gdb.Callback().Update().After("*").Register("buildbuddy:record_metrics", recordMetrics)
+}
+
+func openDB(configurator *config.Configurator, dialect string, connString string) (*gorm.DB, error) {
+	var dialector gorm.Dialector
+	switch dialect {
+	case "sqlite3":
```

Comment:
Are there consts we could use from gorm or that we could define ourselves for these? {"sqlite3", "mysql", "postgres"}?

Reply:
Done.